### PR TITLE
Enumerable Methods Feature 

### PR DESCRIPTION
**Pull request by bouaik**
# Enumerable Methods
## About the Project
In this project i created my own version of Ruby’s enumerable methods.

### CODE REVIEW

File: index.rb [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  def my_each
+    if block_given?
+      0.upto(length - 1) do |i|
+        yield self[i]
+      end
+    else
+      puts 'no block is given'
+    end
+    self
+  end
```
Comment (bolah2009) [inline index.rb]: `my_each` method should return an `Enumerator` when no block is given. 
Check if a block is given at the beginning of the method if no block is given return an `Enumerator` instead else continue with the method logic.

See the difference below:

```ruby
[1,2,3].each
#=> #<Enumerator: [1, 2, 3]:each>

[1,2,3].my_each
no block is given
#=> [1, 2, 3]
```

HINT: Add `return self.to_enum unless block_given?`

### Let's do this for `my_each_with_index` and `my_select`

File: index.rb [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+x.my_each_with_index do |ele, index|
+  puts ele * index
+end
+
+puts '<------ MY SELECT ------>'
+print x.my_select(&:positive?)
+puts
+
+puts '<------ MY ALL ------>'
+words.my_all do |str|
```
Comment (bolah2009) [inline index.rb]: [Optional]

Let remove all these test cases.
You can add it to readme as code examples or/and move it to a different file instead.

File: index.rb [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    value = true
+    if block_given?
+      0.upto(length - 1) do |i|
+        value = false unless yield self[i]
+      end
+    else
+      puts 'no block is given'
+    end
+    value
+  end
```
Comment (bolah2009) [inline index.rb]: #### It is important to note that `my_any?`, `my_all?` and `my_none?` can accept an argument and can also be called without any argument or block. See some cases the current method does not cover below:

 - When a pattern is passed this method should return true if every collection member matches this pattern. Note that pattern could be a `Class` (e.g. `Integer`, `String`) or `RegEx`
   ```ruby
   # When a Class pattern is passed
   irb(main):014:0> [1,2,3].all?(Integer)
   => true
   irb(main):015:0> [1,2,3].my_all(Integer)
   => ArgumentError (wrong number of arguments (given 1, expected 0))

   # When a RegEx pattern is passed
   irb(main):018:0> %w[ant bat cat].all?(/t/)
   => true
   irb(main):019:0> %w[ant bat cat].my_all(/t/)
   => ArgumentError (wrong number of arguments (given 1, expected 0))

   # When no argument is passed
   irb(main):013:0> [nil, true, false].all?
   => false
   irb(main):014:0> [nil, true, false].my_all
   no block is given
   => true
   ```
   Hint: Use ` === ` to check if a member matches a patern.

### Find more details about [`all?` here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F)

### Please let's do this for `my_any?` and `my_none?` too.
_Also, note that methods that return `true` or `false` should end with `?`_

**Review by bolah2009 — CHANGES_REQUESTED** [review]
Hi @bouaik 

Good job with the project so far, however, there is still some issue with some methods (listed below)

See my comments for more details, you should also check [the ruby doc for these methods behaviour](https://ruby-doc.org/core-2.6.5/Enumerable.html). Methods to look out for are the following:

- `my_each`
- `my_each_with_index`
- `my_select`
- `my_all?`
- `my_any?`
- `my_none?`
- `my_map`
- `my_count`
- `my_inject`

Please fix these methods before submitting another CRR.

#### :recycle:  Make these changes and submit another code review request. 👍🏾

#### Feel free to reach me on slack if you having issues implementing these features. :relaxed: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

**Comment by bouaik** [comment]
## Your recommendation has been implemented 
I tried to use === to check the pattern but rubocop suggest to avoid using it.

File: index.rb [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  def my_all?(patt = nil)
+    if block_given?
+      my_each { |i| return false unless yield(i) }
+    elsif patt.nil?
+      my_each { |i| return false unless i }
+    else
+      my_each { |i| return false unless check_patt(i, patt) }
+    end
+    true
+  end
```
Comment (suretrust) [inline index.rb:39]: * [ ] should have working `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_all?(3) != [1,2,3,4,5].all?(3)
```

File: index.rb [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+  def my_any?(patt = nil)
+    if block_given?
+      my_each { |i| return true unless yield(i) }
+    elsif patt.nil?
+      my_each { |i| return true unless i }
+    else
+      my_each { |i| return true unless check_patt(i, patt) }
+    end
+    false
```
Comment (suretrust) [inline index.rb]: * [ ] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5,6].my_any?{ |num| num <= 6 } != [1,2,3,4,5,6].any?{ |num| num <= 6 }

[nil, false, nil, false].my_any? != [nil, false, nil, false].any?

[1,2,3,4,5].my_any?(Integer) != [1,2,3,4,5].any?(Integer)

%w[dog door rod blade].my_any?(Integer) != [dog door rod blade].any?(Integer)

%w[dog door rod blade].my_any?(/z/) != [dog door rod blade].any?(/z/)

%w[dog door rod blade].my_any?('cat') != %w[dog door rod blade].any?('cat')
```

File: index.rb [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      my_each { |i| return false unless i }
+    else
+      my_each { |i| return false unless check_patt(i, patt) }
+    end
+    true
+  end
+
+  def my_none?(arg = nil)
+    !my_all?(arg)
+  end
```
Comment (suretrust) [inline index.rb]: * [ ] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_none?{ |num| num > 5 } != [1,2,3,4,5].none?{ |num| num > 5 }

[nil, false, true, []].my_none? != [nil, false, true, []].none?

[1,2,3,4,5,6].my_none?(String) != [1,2,3,4,5,6].none?(String)

%w[dog door rod blade].my_none?(5) != %w[dog door rod blade].none?(5)
```

File: index.rb [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    count = 0
+    if block_given?
+      0.upto(length - 1) do |i|
+        count += 1 if yield self[i]
+      end
+    else
+      puts 'no block is given'
+    end
+    count
+  end
```
Comment (suretrust) [inline index.rb]: * [ ] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_count != [1,2,3,4,5].count

[1,2,3,4,5].my_count(1) != [1,2,3,4,5].count(1)
```

File: index.rb [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+      else
+        puts 'no block is given'
+      end
+    else
+      0.upto(length - 1) do |i|
+        array << proc.call(i)
+      end
+    end
+    array
+  end
```
Comment (suretrust) [inline index.rb]: * [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_map != [1,2,3,4,5].map
```

File: index.rb [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    end
+    if block_given?
+      index.upto(length - 1) do |i|
+        acc = yield(acc, self[i])
+      end
+    else
+      puts 'no block is given'
+    end
+    acc
+  end
```
Comment (suretrust) [inline index.rb]: * [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
5..10.my_inject(4) { |prod, n| prod * n } != 5..10.inject(4) { |prod, n| prod * n }

[1,2,3,4,5].my_inject(:+) != [1,2,3,4,5].inject(:+)
```

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi @bouaik, 👋 

### Great job working on the suggested changes. 🙌 However, we have a few test cases to go. 👇 

**Kindly submit for another code review request when these have been fixed.** _Happy coding!_ 🎉

File: index.rb [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    end
+  end
+
+  def check_patt(idx, patt)
+    if patt.is_a? Class
+      idx.is_a?(patt)
+    elsif patt.is_a? Regexp
+      patt.match?(idx)
+    end
+  end
```
Comment (MauricioRobayo) [inline index.rb]: Nice! but missing the case when it is not a Regexp nor a Class, for example:

```ruby
    else
      patt == idx
```

File: index.rb [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+    end
+    count
+  end
+
+  def my_map(proc = nil)
+    array = []
+    if proc.nil?
+      return to_enum(:my_map) unless block_given?
+
+      my_each { |i| array << yield(self[i]) }
```
Comment (MauricioRobayo) [inline index.rb]: 🤔Seem to be an issue here. Isn't `i` already the element of the array? Why access `self[i]`?

**Review by MauricioRobayo — CHANGES_REQUESTED** [review]
Well done fixing some of the issues, almost there!

Please double-check `my_all?`, `my_any?`, and `my_none?` when a pattern other than a RegEx or a Class is given, for example:

```ruby
p [1].my_all?(1) == [1].all?(1)
p [1].my_any?(1) == [1].any?(1)
p [1].my_none?(1) == [1].none?(1)
```

Also, double-check `my_map`, there is an issue, for example, this throws an error:

```ruby
[1, 2].map { |num| num < 2 }
``` 
TL;DR:

* [ ] Fix `my_all?` when a pattern other than a RegEx or a Class is given.
* [ ] Fix `my_any?` when a pattern other than a RegEx or a Class is given.
* [ ] Fix `my_none?` when a pattern other than a RegEx or a Class is given.
* [ ] Fix `my_map` when running a block.


Status: __Request changes__.

**Review by kobiyoyo — APPROVED** [review]
# Project Approved :thumbsup:
## Hello @bouaik ,  You have done a great job with this project. :clap:
